pkg/types/codec.go

```diff
@@ -0,0 +1,39 @@
+package types
+
+import (
+	"context"
+
+	"github.com/smartcontractkit/libocr/offchainreporting2plus/types"
+)
+
+type Encoder interface {
+	Encode(ctx context.Context, item any, itemType string) (types.Report, error)
+	// GetMaxEncodingSize returns the max size in bytes if n elements are supplied for all top level dynamically sized elements.
+	// If no elements are dynamically sized, the returned value will be the same for all n.
+	// If there are multiple levels of dynamically sized elements, or itemType cannot be found,
+	// InvalidTypeError will be returned.
+	GetMaxEncodingSize(ctx context.Context, n int, itemType string) (int, error)
+}
+
+type Decoder interface {
+	Decode(ctx context.Context, raw []byte, into any, itemType string) error
```

Comment:
Should this take a `types.Report` for symmetry?
```suggestion
	Decode(ctx context.Context, raw types.Report, into any, itemType string) error
```

Reply:
But the decoder won't strictly decode reports

Reply:
Then should the encoder be producing `[]byte`? Or am I misunderstanding something?

Reply:
I think so, if I understood correctly yesterday. This will be used both for chain reader and for Report Codec. So it should be just bytes